ChronoVisor/include/ChronicleMetaDirectory.h

```diff
@@ -19,23 +19,23 @@ class ChronicleMetaDirectory {
 //    std::shared_ptr<std::unordered_map<std::string, Chronicle *>> getChronicleMap() { return chronicleMap_; }
     std::unordered_map<uint64_t, Chronicle *> *getChronicleMap() { return chronicleMap_; }
 
-    bool create_chronicle(const std::string& name);
-    bool create_chronicle(const std::string& name, const std::unordered_map<std::string, std::string>& attrs);
-    bool destroy_chronicle(const std::string& name, int flags);
-    uint64_t acquire_chronicle(const std::string& name, int flags);
-    bool release_chronicle(const std::string& name, int flags);
-
-    bool create_story(std::string &chronicle_name, const std::string& story_name, const std::unordered_map<std::string, std::string>& attrs);
-    bool destroy_story(std::string &chronicle_name, const std::string& story_name, int flags);
-    std::vector<uint64_t> get_story_list(std::string &chronicle_name);
-    uint64_t acquire_story(const std::string &chronicle_name, const std::string& story_name, int flags);
-    bool release_story(const std::string &chronicle_name, const std::string &story_name, int flags);
+    int create_chronicle(const std::string& name);
+    int create_chronicle(const std::string& name, const std::unordered_map<std::string, std::string>& attrs);
+    int destroy_chronicle(const std::string& name, int& flags);
+    int acquire_chronicle(const std::string& name, int& flags);
+    int release_chronicle(const std::string& name, int& flags);
+
+    int create_story(std::string &chronicle_name, const std::string& story_name, const std::unordered_map<std::string, std::string>& attrs);
+    int destroy_story(std::string &chronicle_name, const std::string& story_name, int& flags);
+    int get_story_list(std::string &chronicle_name, std::vector<std::string> &story_name_list);
+    int acquire_story(const std::string& chronicle_name, const std::string& story_name, int& flags);
+    int release_story(const std::string& chronicle_name, const std::string& story_name, int& flags);
 
     uint64_t record_event(uint64_t sid, void *data);
     uint64_t playback_event(uint64_t sid);
 
-    std::string get_chronicle_attr(std::string &name, const std::string& key);
-    bool edit_chronicle_attr(std::string &name, const std::string& key, std::string value);
+    int get_chronicle_attr(std::string& name, const std::string& key, std::string& value);
+    int edit_chronicle_attr(std::string& name, const std::string& key, const std::string& value);
```

Comment:
Do you want to duplicate the data in chronicleMap and StoryMaps and maintain acquiredChronicleMap_ and acquiredStoryMap? Can you store the acquired info in the primary data structures? Otherwise, you will have to update the primary and auxiliary data structures atomically (within the same critical section).

Reply:
Currently, both ChroniMap and acquiredChronicleMap are both unordered_map<uint64_t, Chronicle *>. Their values are pointers to the same Chronicle instances. They are like two access points to the objects. We might want to maintain something else in acquiredChronicleMap, for instance, information of a Chronicle that only effective for this acquisition session. I think it is okay for now. We only need to make one change via either map. There is no primary or auxilliary data structures.

Reply:
Won't the values pointed to by the pointers be updated when the primary data structure inserts/deletes?

Reply:
Yes, ChronicleMap and acquiredChronicleMap are kinda primary/auxiliary from your point of view. I wanted to say that we do not maintain two copies of each entry. I do have a simple single-threaded test case ready which emulates this situation. At least the already-inserted entry does not change when a new entry is inserted (not overwrite the existing one). The data race could be a problem. What I believe we can do is to make sure there is no concurrent access to the same Chronicle/Story. I will update my test case to multithreaded version to test data race.

Reply:
Yeah, if we have a global mutex, which is the situation now, this may work. But there will be data races when you replace the stl maps and global mutex with concurrent hash tables. Maybe you should add this to our design document and we don't forget this issue when we replace the stl maps with HCL containers.